### PR TITLE
feat(matchers): add toContainClass

### LINKED ISSUE #14117: [Feature] add expect(locator).toContainClass()
Hi there)

I have a strange behavior with toHaveClass assertion.

I have an html

```html
<label class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-error css-1958la6-MuiFormLabel-root-MuiInputLabel-root" data-shrink="false" for="mui-1" id="mui-1-label"></label>
```

And I have assertion:

```tsx
await expect(label).toHaveClass('Mui-error')
```

which fails, but this one is good:

```tsx
await expect(label).toHaveClass('MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-error css-1958la6-MuiFormLabel-root-MuiInputLabel-root')
```

Is that intended?

I expected `toHaveClass` be like: "Hey, label, do you have class Mui-error in your bag?", but instead I have "Hey, label, do you have exactly one class and is it Mui-error?".

### PR DESCRIPTION
**Pull request by mxschmitt**
Fixes https://github.com/microsoft/playwright/issues/14117
Fixes https://github.com/microsoft/playwright/pull/14219

### CODE REVIEW

File: packages/playwright-core/src/server/frames.ts [hunk 1]
```diff
@@ -1345,11 +1345,10 @@ export class Frame extends SdkObject {
       omitAttached = true;
     else if (options.isNot && options.expression === 'to.have.count' && options.expectedNumber !== 0)
       omitAttached = true;
-    else if (!options.isNot && options.expression.endsWith('.array') && options.expectedText!.length === 0)
+    else if (!options.isNot && options.expression.endsWith('.array') && options.expectedText?.length === 0)
```
Comment (dgozman) [inline packages/playwright-core/src/server/frames.ts:1348]: I think this does not work with `contain.class.array` because `undefined !== 0`, but it should.

File: packages/playwright-core/src/server/injected/injectedScript.ts [hunk 2]
```diff
@@ -1107,6 +1112,15 @@ export class InjectedScript {
       return { received, matches };
     }
 
+    if (expression === 'to.contain.class.array') {
+      if (!Array.isArray(options.expectedValue) || options.expectedValue.length !== elements.length)
```
Comment (dgozman) [inline packages/playwright-core/src/server/injected/injectedScript.ts:1116]: How can `expectedValue` be not an array?

File: packages/playwright-test/src/matchers/matchers.ts [hunk 3]
```diff
@@ -164,6 +164,23 @@ export function toHaveClass(
   }
 }
 
+export function toContainClass(
+  this: ReturnType<Expect['getState']>,
+  locator: LocatorEx,
+  expected: string | string[],
+  options?: { timeout?: number },
```
Comment (dgozman) [inline packages/playwright-test/src/matchers/matchers.ts:171]: `ignoreCase` right away?

File: docs/src/api/class-locatorassertions.md [hunk 4]
```diff
@@ -721,6 +721,93 @@ await Expect(locator).ToBeVisibleAsync();
 ### option: LocatorAssertions.toBeVisible.timeout = %%-csharp-java-python-assertions-timeout-%%
 * since: v1.18
 
+## async method: LocatorAssertions.toContainClass
+* since: v1.24
+* langs:
+  - alias-java: containsClass
+
+Ensures the [Locator] points to an element that contains the given CSS class (or multiple).
+In contrast to [`method: LocatorAssertions.toHaveClass`] which requires that the [Locator] has exactly the provided classes, `toContainClass` verifies that the [Locator] has a subset (or all) the given CSS classes.
```
Comment (dgozman) [inline docs/src/api/class-locatorassertions.md:730]:
```suggestion
In contrast to [`method: LocatorAssertions.toHaveClass`] which requires that the [Locator] has exactly the provided classes, `toContainClass` verifies that the [Locator] has a subset (or all) of the given CSS classes.
```

File: packages/playwright-test/src/matchers/matchers.ts (hunk 3 above)
Comment (mxschmitt) [reply packages/playwright-test/src/matchers/matchers.ts:171]: Not sure if this would worth doing it. But implementation wise it would mean, that I should use `expectedText` instead of `expectedValue`  to pass the `ignoreCase` value over like we do in the other places.

File: packages/playwright-core/src/server/injected/injectedScript.ts [hunk 5]
```diff
@@ -1270,6 +1276,18 @@ class ExpectedTextMatcher {
...
+  private matchesClassList(received: string): boolean {
+    const expected = this.splitClassList(this._string || '');
+    const normalizedReceived = this.splitClassList(received).map(r => this.normalize(r));
+    return expected.every(classListEntry =>
+      normalizedReceived.includes(this.normalize(classListEntry))
+    );
+  }
+
+  private splitClassList(classList: string): string[] {
+    return classList.split(/\s+/g);
```
Comment (dgozman) [inline packages/playwright-core/src/server/injected/injectedScript.ts:1288]: Should you trim first?

File: packages/playwright-core/src/server/injected/injectedScript.ts [hunk 6]
```diff
@@ -1270,6 +1276,18 @@ class ExpectedTextMatcher {
...
 
+  private matchesClassList(received: string): boolean {
+    const expected = this.splitClassList(this._string || '');
+    const normalizedReceived = this.splitClassList(received).map(r => this.normalize(r));
+    return expected.every(classListEntry =>
+      normalizedReceived.includes(this.normalize(classListEntry))
+    );
+  }
+
+  private splitClassList(classList: string): string[] {
```
Comment (dgozman) [inline packages/playwright-core/src/server/injected/injectedScript.ts:1287]: Might as well call this `normalizeClassList` and split+normalize inside.

File: tests/playwright-test/playwright.expect.misc.spec.ts [hunk 7]
```diff
@@ -262,6 +262,56 @@ test('should support toHaveClass w/ array', async ({ runInlineTest }) => {
...
+            <span class="alice item baz"></span>
+            <span class="bob item baz"></span>
+          </div>
+        \`);
+        const locator = page.locator('div');
+        await expect(locator).toContainClass('foo');
+        await expect(locator).toContainClass('bar');
+        await expect(locator).toContainClass('baz');
+        await expect(locator).toContainClass('foo baz');
+        await expect(locator).toContainClass('baz foo');
```
Comment (dgozman) [inline tests/playwright-test/playwright.expect.misc.spec.ts:282]: - Add a test with leading/trailing whitespace in `expected`.
- Add a test with empty `expected`.

**Review by dgozman — APPROVED** [review]
